src/UI/Layout.hs

```diff
@@ -25,66 +29,80 @@ offset by = wrap . Offset by
 stack = unStack . foldMap Stack
 
 measureLayout :: Real a => Layout a (Size a) -> Size a
-measureLayout = fromMaybe (Size 0 0) . fitLayoutTo (pure Nothing)
+measureLayout = fromMaybe (Size 0 0) . fitLayout (pure Nothing)
+
+fitLayout :: Real a => Size (Maybe a) -> Layout a (Size a) -> Maybe (Size a)
+fitLayout = fitLayoutWith layoutSizeAlgebra
 
-fitLayoutTo :: Real a => Size (Maybe a) -> Layout a (Size a) -> Maybe (Size a)
-fitLayoutTo maxSize layout = case fromF layout of
+fitLayoutAndAnnotate :: Real a => Size (Maybe a) -> Layout a (Size a) -> ALayout a (Size a) (Maybe (Size a))
+fitLayoutAndAnnotate = fitLayoutWith (annotatingBidi layoutSizeAlgebra)
+
+layoutSizeAlgebra :: Real a => CofreerF (FreerF (LayoutF a) (Size a)) (Size (Maybe a)) (Maybe (Size a)) -> Maybe (Size a)
+layoutSizeAlgebra (Cofree maxSize runC layout) = case layout of
   Pure size | maxSize `encloses` size -> Just (fromMaybe <$> size <*> maxSize)
-  Free (Inset inset rest) | maxSize `encloses` (2 * inset) -> (2 * inset +) <$> fitLayoutTo (subtractSize (2 * inset)) (toF rest)
-  Free (Offset offset rest) | maxSize `encloses` pointSize offset -> (pointSize offset +) <$> fitLayoutTo (subtractSize (pointSize offset)) (toF rest)
-  Free (Resizeable resize) -> fitLayoutTo maxSize (toF (resize maxSize))
+  Free runF l -> case l of
+    Inset by child -> (2 * by +) <$> runC (runF child)
+    Offset by child -> (pointSize by +) <$> runC (runF child)
+    Resizeable resize -> runC (runF (resize maxSize))
+    Measure child withMeasurement -> runC (runF child) >>= runC . runF . withMeasurement
   _ -> Nothing
   where maxSize `encloses` size = and (maybe (const True) (>=) <$> maxSize <*> size)
-        subtractSize size = liftA2 (-) <$> maxSize <*> (Just <$> size)
 
 
--- Geometry
+fitLayoutWith :: Real a => (CofreerF (FreerF (LayoutF a) (Size a)) (Size (Maybe a)) b -> b) -> Size (Maybe a) -> Layout a (Size a) -> b
+fitLayoutWith = curry . (`hylo` coalgebra)
+  where coalgebra :: Real a => (Size (Maybe a), Layout a (Size a)) -> CofreerF (FreerF (LayoutF a) (Size a)) (Size (Maybe a)) (Size (Maybe a), Layout a (Size a))
+        coalgebra (maxSize, layout) = Cofree maxSize id $ case runFreer layout of
+          Pure size -> Pure size
+          Free run l -> Free id $ case l of
+            Inset by child -> Inset by (subtractSize maxSize (2 * by), run child)
+            Offset by child -> Offset by (subtractSize maxSize (pointSize by), run child)
+            Resizeable resize -> Resizeable ((,) maxSize . run . resize)
+            Measure child withMeasurement -> Measure (maxSize, run child) ((,) maxSize . run . withMeasurement)
 
-data Rect a = Rect { origin :: !(Point a), size :: !(Size a) }
-  deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
-data Point a = Point { x :: !a, y :: !a }
-  deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
+        subtractSize maxSize size = liftA2 (-) <$> maxSize <*> (Just <$> size)
```

Comment:
Fitting the layout (narrowing the max size according to the terms of the layout functor) is completely orthogonal to the produced result. By taking advantage of that fact, we’re able to reuse the same fitting logic for multiple different computations on layouts.